samples/hello_world/src/main.c

```diff
@@ -4,7 +4,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#include <zephyr.h>
+#include <autoconf.h>
```

Comment:
What motivated this change?

Using `zephyr.h` as a sort of "go-to" header, rather than identifying and manually specifying the required individual headers, is much simpler and less confusing to the beginners; in fact, that is why this header exists in the first place.

Reply:
Ref.

https://github.com/zephyrproject-rtos/zephyr/issues/41543

`<zephyr.h>` in particular is confusing, as it's just `<kernel.h>` + `__ZEPHYR__`. It's like a `<Windows.h>`. Btw, I think it would be nice to have all public headers in `zephyr/` so that we could have `<zephyr/kernel.h>`.

Reply:
Btw, look at what we pull with <kernel.h> that we don't need:

![image](https://user-images.githubusercontent.com/25011557/157530102-f39bd621-a3c5-4a24-8e66-32ac797b0b1e.png)

:upside_down_face:

Reply:
Well... someone needs it.  Zephyr leans heavily on header inlining (LTO is always just over the horizon, but right now inlined code is our tool), especially at the arch layer, and all those seemingly internal headers are needed to expose implementations to the APIs apps "need".

I think there's an excellent argument to be made that a(nother) cleanup pass on our header tangle is probably due.  And there's at least a vaguely justifiable bikeshed discussion to be had[1] about whether a "catchall" header for "the whole Zephyr API" should exist, whether it should be named "zephyr.h" or "kernel.h" (and why they are separate files at all), and whether or not code in samples/ and tests/ should be using it.

But I mean, **if** we have such a header, clearly we'd want to be using it in hello_world, the basic archetype of "A Zephyr App".

[1] Not by me!

(FWIW: this bug was totally worth it just to see that graph)

Reply:
> Well... someone needs it.

Maybe kernel.h itself, but not the `main.c` in hello_world sample.

> I think there's an excellent argument to be made that a(nother) cleanup pass on our header tangle is probably due. And there's at least a vaguely justifiable bikeshed discussion to be had[1] about whether a "catchall" header for "the whole Zephyr API" should exist, whether it should be named "zephyr.h" or "kernel.h" (and why they are separate files at all), and whether or not code in samples/ and tests/ should be using it.

Any inputs on https://github.com/zephyrproject-rtos/zephyr/issues/41543 are welcome :-)